### PR TITLE
NoticeTemplateReporter helpers

### PR DESCRIPTION
**Pull request by mnonnenmacher**
See the commit messages for details.

### CODE REVIEW

File: reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt [hunk 1]
```diff
@@ -177,6 +177,16 @@ class NoticeTemplateReporter : Reporter {
                 licenseConfiguration.licenses.find { it.id == resolvedLicense.license }?.includeInNoticeFile ?: true
             }
 
+        /**
+         * Filter licenses that are configured to require a
+         * [source code offer in the notice file][License.includeInNoticeFile] in the [LicenseConfiguration].
+         */
+        @Suppress("UNUSED") // This function is used in the templates.
+        fun filterIncludeSourceCodeOfferInNoticeFile(licenses: List<ResolvedLicense>): List<ResolvedLicense> =
```
Comment (fviernau) [inline reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt:199]: Please use `Collection<ResolvedLicense>` as parameter.

File: reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt [hunk 2]
```diff
@@ -177,6 +177,16 @@ class NoticeTemplateReporter : Reporter {
                 licenseConfiguration.licenses.find { it.id == resolvedLicense.license }?.includeInNoticeFile ?: true
             }
 
+        /**
+         * Filter licenses that are configured to require a
+         * [source code offer in the notice file][License.includeInNoticeFile] in the [LicenseConfiguration].
+         */
+        @Suppress("UNUSED") // This function is used in the templates.
```
Comment (fviernau) [inline reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt:198]: Would it make sense to make this an extension function of `List<ResolvedLicense>`?

File: reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt [hunk 3]
```diff
@@ -166,7 +167,20 @@ class NoticeTemplateReporter : Reporter {
...
      * A collection of helper functions for the Freemarker templates.
      */
-    class NoticeHelper(private val licenseConfiguration: LicenseConfiguration) {
+    class NoticeHelper(private val ortResult: OrtResult, private val licenseConfiguration: LicenseConfiguration) {
+        /**
+         * Filter [packages] that are a dependency of at least one of the provided [projects][projectIds].
+         */
+        @Suppress("UNUSED") // This function is used in the templates.
+        fun filterByProjects(
+            packages: List<PackageNoticeModel>,
```
Comment (fviernau) [inline reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt:176]: Please use `Collection` for the function parameters.

File: reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt [hunk 4]
```diff
@@ -166,7 +167,20 @@ class NoticeTemplateReporter : Reporter {
...
+            projectIds: List<Identifier>
+        ): List<PackageNoticeModel> {
+            val dependencies = projectIds.mapNotNull { ortResult.getProject(it) }
+                .flatMapTo(mutableSetOf()) { it.collectDependencies() }
+            return packages.filter { pkg -> pkg.id in dependencies }
```
Comment (fviernau) [inline reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt:181]: Consider empty line before this one.

File: reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt [hunk 5]
```diff
@@ -166,7 +167,20 @@ class NoticeTemplateReporter : Reporter {
     /**
      * A collection of helper functions for the Freemarker templates.
      */
-    class NoticeHelper(private val licenseConfiguration: LicenseConfiguration) {
+    class NoticeHelper(private val ortResult: OrtResult, private val licenseConfiguration: LicenseConfiguration) {
+        /**
+         * Filter [packages] that are a dependency of at least one of the provided [projects][projectIds].
```
Comment (fviernau) [inline reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt:172]: Turning the preceeding "Filter" into "Return" would IMO be more clear about the filter criteria. 
As-is it is easier to understand it as the inversion of the criteria IMO.

File: reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt [hunk 6]
```diff
@@ -184,7 +198,8 @@ class NoticeTemplateReporter : Reporter {
         @Suppress("UNUSED") // This function is used in the templates.
         fun filterIncludeSourceCodeOfferInNoticeFile(licenses: List<ResolvedLicense>): List<ResolvedLicense> =
             licenses.filter { resolvedLicense ->
-                licenseConfiguration.licenses.find { it.id == resolvedLicense.license }?.includeSourceCodeOfferInNoticeFile ?: false
+                licenseConfiguration.licenses.find { it.id == resolvedLicense.license }?.includeSourceCodeOfferInNoticeFile
+                    ?: false
```
Comment (fviernau) [inline reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt:202]: not part of this commit?!

**Review by fviernau — CHANGES_REQUESTED** [review]

File: reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt (hunk 6 above)
Comment (mnonnenmacher) [reply reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt]: Done.

File: reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt (hunk 5 above)
Comment (mnonnenmacher) [reply reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt]: Done.

File: reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt (hunk 4 above)
Comment (mnonnenmacher) [reply reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt:183]: Done.

File: reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt (hunk 3 above)
Comment (mnonnenmacher) [reply reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt]: Done.

File: reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt (hunk 1 above)
Comment (mnonnenmacher) [reply reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt]: Done.

File: reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt (hunk 2 above)
Comment (mnonnenmacher) [reply reporter/src/main/kotlin/reporters/NoticeTemplateReporter.kt:201]: No, that would make the function harder to call from the template, because it does not support Kotlin extension functions.

**Review by fviernau — APPROVED** [review]
